lib/src/rules/use_build_context_synchronously.dart

```diff
@@ -0,0 +1,199 @@
+// Copyright (c) 2021, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/type.dart';
+
+import '../analyzer.dart';
+
+// todo (pq): flesh out storing context gotchas -- codify here or in another lint?
+
+const _desc = r'Do not use BuildContexts across async calls.';
+
+const _details = r'''
+**DO NOT** use BuildContexts across async calls.
+
+TODO: add rationale.
+
+TODO: describe mechanics.
+
+
+**BAD:**
+```
+class MyState extends State<MyWidget> {
+  void m() async {
+    // Uses context from State.
+    Navigator.of(context).pushNamed('routeName');
+
+    await Future<void>.delayed(Duration());
+    // ^--- async gap.
+
+    // Without a mounted check, this is unsafe.
+    Navigator.of(context).pushNamed('routeName'); // LINT
+  }
+}
+```
+
+**GOOD:**
+```
+class MyState extends State<MyWidget> {
+  void m() async {
+    // Uses context from State.
+    Navigator.of(context).pushNamed('routeName');
+
+    await Future<void>.delayed(Duration());
+
+    if (!mounted) return;
+
+    // OK. We checked mounted checked first.
+    Navigator.of(context).pushNamed('routeName'); // OK
+  }
+}
+```
+''';
+
+class UseBuildContextSynchronously extends LintRule implements NodeLintRule {
+  UseBuildContextSynchronously()
+      : super(
+            name: 'use_build_context_synchronously',
+            description: _desc,
+            details: _details,
+            group: Group.errors,
+            maturity: Maturity.experimental);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    final visitor = _Visitor(this);
+    registry.addMethodInvocation(this, visitor);
+    registry.addInstanceCreationExpression(this, visitor);
+    registry.addFunctionExpressionInvocation(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor {
+  static const _nameBuildContext = 'BuildContext';
+  final Uri _uriFramework;
+
+  // todo (pq): replace in favor of flutter_utils.isBuildContext
```

Comment:
This comment probably became detached after a sort.

Reply:
yep.  We should really fix this some day!